### PR TITLE
feat(msw): split polymorphic mocks into its own functions

### LINKED ISSUE #1318: Feature request: Split polymorphic mocks into it's own functions
## What are the steps to reproduce this issue?

Using the following schema:
```yml
paths:
  /api/sample:
    get:
      summary: sample
      responses:
        '200':
          description: Success
          content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/TypeObj1'
                  - $ref: '#/components/schemas/TypeObj2'
components:
  schemas:
    Type:
      enum:
        - Type1
        - Type2
      type: string
    Base:
      type: object
      required:
        - type
      properties:
        type:
          allOf:
            - $ref: '#/components/schemas/Type'
      discriminator:
        propertyName: type
        mapping:
          Type1: '#/components/schemas/TypeObj1'
          Type2: '#/components/schemas/TypeObj2'
    TypeObj1:
      type: object
      allOf:
        - $ref: '#/components/schemas/Base'
      required:
        - type
      properties:
        moreProp:
          type: string
    TypeObj2:
      type: object
      allOf:
        - $ref: '#/components/schemas/Base'
      required:
        - type
      properties:
        prop:
          type: string
```
with `mock: true` in the configuration

## What happens?

Generated mock is 
```ts
export const getGetApiSampleResponseMock = (
  overrideResponse: any = {},
): GetApiSample200 =>
  faker.helpers.arrayElement([
    {
      ...overrideResponse,
      moreProp: faker.helpers.arrayElement([faker.word.sample(), undefined]),
      type: faker.helpers.arrayElement(['Type1'] as const),
      ...overrideResponse,
    },
    {
      ...overrideResponse,
      prop: faker.helpers.arrayElement([faker.word.sample(), undefined]),
      type: faker.helpers.arrayElement(['Type2'] as const),
      ...overrideResponse,
    },
  ]);
```

## What were you expecting to happen?

I'd like the mock to be split up so I can reference a specific type, something like this:
```ts
export const getTypeObj1Mock = (overrideResponse: any ={}): TypeObj1 => ({
      moreProp: faker.helpers.arrayElement([faker.word.sample(), undefined]),
      type: faker.helpers.arrayElement(['Type1'] as const),
      ...overrideResponse,
});
export const getTypeObj2Mock = (overrideResponse: any ={}): TypeObj2 => ({
      prop: faker.helpers.arrayElement([faker.word.sample(), undefined]),
      type: faker.helpers.arrayElement(['Type2'] as const),
      ...overrideResponse,
});
export const getGetApiSampleResponseMock = (
  overrideResponse: any = {},
): GetApiSample200 =>
  faker.helpers.arrayElement([
    getTypeObj1Mock(overrideResponse),
    getTypeObj2Mock(overrideResponse),
  ]);
```

## Any other comments?

This would help when testing, so I can test that when the api returns a specific type, the correct thing gets rendered.

## What versions are you using?

**Package Version:** 6.27.1

### PR DESCRIPTION
**Pull request by AffeJonsson**
## Status

**READY**

## Description

Fixes #1318 
This splits response mocks where the response contains objects that have a base object.
The specification in the linked issue is an example of a response mock that will get split.
image

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Changelog Entry (unreleased)

## Steps to Test or Reproduce

1. Generate using a specification that contains a `allOf` prop, e.g. the one in the linked issue. 
2. Notice that the response mock contains calls to functions that has been generated above.
3. Notice that no errors exists

### CODE REVIEW

**Comment by soartec-lab** [comment]
@AffeJonsson 

A conflict occurred when I merged the following PR. Could you please check again and correct it?

https://github.com/anymaniax/orval/pulls?q=is%3Apr+is%3Aclosed

**Comment by AffeJonsson** [comment]
> @AffeJonsson
> 
> A conflict occurred when I merged the following PR. Could you please check again and correct it?
> 
> https://github.com/anymaniax/orval/pulls?q=is%3Apr+is%3Aclosed

Merged master and did a small update to make overrideResponse param Partial<name>, just like the new overrideResponse code

**Review by melloware — APPROVED** [review]

**Comment by AffeJonsson** [comment]
Maybe wait with this until #1199 is complete? I assume we might need to change this when it is completed anyway

**Comment by soartec-lab** [comment]
@AffeJonsson 

No, they says it will take time, so we unnecessary wait. And. I'll share it to they.
I would like to check this PR, but please forgive me if it takes a little more time.

**Comment by karlismelderis-mckinsey** [comment]
it would make sense to regenerate samples

example in screenshot has `overrideResponse` twice 🤔
```
...overrideResponse
// some faker
...overrideResponse
```

**Comment by AffeJonsson** [comment]
> it would make sense to regenerate samples
> 
> example in screenshot has `overrideResponse` twice 🤔
> 
> ```
> ...overrideResponse
> // some faker
> ...overrideResponse
> ```

This is no longer the case after your PR was merged, only one overrideResponse is now present

File: packages/mock/src/faker/resolvers/value.ts [hunk 1]
```diff
@@ -101,8 +105,24 @@ export const resolveMockValue = ({
       },
       imports,
       existingReferencedProperties,
+      functions,
       allowOverride,
     });
+    if (newSchema.allOf) {
```
Comment (soartec-lab) [inline packages/mock/src/faker/resolvers/value.ts]: Not only can objects be separated using `oneOf`, but also objects can now be separated using `allOf`? ?

File: packages/mock/src/types.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ export interface MockDefinition {
   name: string;
   overrided?: boolean;
   includedProperties?: string[];
+  functions?: string[];
```
Comment (soartec-lab) [inline packages/mock/src/types.ts]: I would like to rename `functions` for a couple of reasons.

1. It is difficult to tell whether it is a list of typescript `Function` or a collection of function definitions as a string.
2. `MockDefinition` already has a field `value` to get the value, so we need to clarify what value this field is.

If it's functions of split mock definition, how about `splitMockImplementions`?

**Review by soartec-lab — COMMENTED** [review]
@AffeJonsson 
Thank you for great update. I check it and commented.

File: packages/mock/src/faker/resolvers/value.ts (hunk 1 above)
Comment (AllieJonsson) [reply packages/mock/src/faker/resolvers/value.ts]: My mistake, I did this backwards. I have now changed this to look at the `combine.separator === 'oneOf` instead, which is much better.

File: packages/mock/src/types.ts (hunk 2 above)
Comment (AllieJonsson) [reply packages/mock/src/types.ts]: Good suggestion, done!

**Comment by AffeJonsson** [comment]
Seems I broke tests with this, fixing...

File: packages/mock/src/msw/index.ts [hunk 3]
```diff
@@ -90,10 +93,18 @@ const generateDefinition = (
...
   const handlerName = `${handlerNameBase}${pascal(name)}`;
 
-  const mockImplementation = isReturnHttpResponse
-    ? `export const ${getResponseMockFunctionName} = (${isResponseOverridable ? `overrideResponse: Partial< ${returnType} > = {}` : ''})${mockData ? '' : `: ${returnType}`} => (${value})\n\n`
+  const addedSplitMockImplementations = splitMockImplementations.slice(
+    oldSplitMockImplementations.length,
+  );
+  splitMockImplementations.push(...addedSplitMockImplementations);
+  const mockImplementations = addedSplitMockImplementations.length
+    ? `//#region ${getResponseMockFunctionName} sub functions\n${addedSplitMockImplementations.join('\n\n')}\n//#endregion\n\n`
```
Comment (soartec-lab) [inline packages/mock/src/msw/index.ts]: If you want to include annotate, I think another way to express them using variable names would be appropriate. Or i think you can just delete it if it's not necessary.

File: packages/mock/src/msw/mocks.ts [hunk 4]
```diff
@@ -133,7 +134,9 @@ export const getResponsesMockDefinition = ({
   transformer?: (value: unknown, definition: string) => string;
   context: ContextSpecs;
   mockOptions?: GlobalMockOptions;
+  splitMockImplementations: string[];
 }) => {
+  // console.log(responses);
```
Comment (soartec-lab) [inline packages/mock/src/msw/mocks.ts]: Looks like I forgot to turn this off.

**Review by soartec-lab — COMMENTED** [review]
@AffeJonsson 
I checked again and requested a minor correction. Please check.

**Comment by AffeJonsson** [comment]
> @AffeJonsson Please let me know when this is ready for me to review 👍

It's ready!

**Review by soartec-lab — APPROVED** [review]
Thanks!

**Pull request merged**